### PR TITLE
PWGHF:Add Xicc analysis

### PR DESCRIPTION
**Pull request by ginnocen**

### CODE REVIEW

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{
+    adaptAnalysisTask<HFCandidateCreatorXicc>(cfgc, TaskName{"hf-cand-creator-xicc"}),
+    adaptAnalysisTask<HFCandidateCreatorXiccExpressions>(cfgc, TaskName{"hf-cand-creator-xicc-expressions"})};
+  const bool doMC = cfgc.options().get<bool>("doMC");
+  if (doMC) {
+    workflow.push_back(adaptAnalysisTask<HFCandidateCreatorXiccMC>(cfgc, TaskName{"hf-cand-creator-xicc-mc"}));
+  }
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Name the tasks such that `TaskName` is not necessary.

File: Analysis/Tasks/PWGHF/taskXicc.cxx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+     {"hPtRecBg", "#Xi^{++}_{cc} candidates (rec. unmatched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGen", "#Xi^{++}_{cc} candidates (gen. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGenSig", "#Xi^{++}_{cc} candidates (rec. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hEtaRecSig", "#Xi^{++}_{cc} candidates (rec. matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hEtaRecBg", "#Xi^{++}_{cc} candidates (rec. unmatched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hYRecSig", "#Xi^{++}_{cc} candidates (rec. matched);rapidity;entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hYRecBg", "#Xi^{++}_{cc} candidates (rec. unmatched);rapidity;entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hEtaGen", "#Xi^{++}_{cc} candidates (gen. matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hYGen", "#Xi^{++}_{cc} candidates (gen. matched);rapidity;entries", {HistType::kTH1F, {{100, -2., 2.}}}}}};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskXicc.cxx]: Fix the histogram titles according to whether a histogram is filled with the sample of candidates or particles.

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h [hunk 3]
```diff
@@ -586,10 +586,12 @@ auto InvMassXicToPiKP(const T& candidate)
 {
   return InvMassLcpiKp(candidate);
 }
+
 } // namespace hf_cand_prong3
 
 // 3-prong decay candidate table
 DECLARE_SOA_TABLE(HfCandProng3Base, "AOD", "HFCANDP3BASE", //!
+                  o2::soa::Index<>,
```
Comment (vkucera) [inline Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:589]: What does this do and why is it needed?

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h [hunk 4]
```diff
@@ -644,6 +646,96 @@ DECLARE_SOA_TABLE(HfCandProng3MCGen, "AOD", "HFCANDP3MCGEN", //!
...
+namespace hf_cand_xicc
+{
+DECLARE_SOA_INDEX_COLUMN_FULL(Index0, index0, int, HfCandProng3, "_0"); // Xic index
+// MC matching result:
+DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, int8_t);         // reconstruction level
+DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t);         // generator level
+DECLARE_SOA_COLUMN(OriginMCRec, originMCRec, int8_t);               // particle origin, reconstruction level
+DECLARE_SOA_COLUMN(OriginMCGen, originMCGen, int8_t);               // particle origin, generator level
+DECLARE_SOA_COLUMN(FlagMCDecayChanRec, flagMCDecayChanRec, int8_t); // resonant decay channel flag, reconstruction level
+DECLARE_SOA_COLUMN(FlagMCDecayChanGen, flagMCDecayChanGen, int8_t); // resonant decay channel flag, generator level
```
Comment (vkucera) [inline Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h]: Are these needed?

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                                  xicCand.index2_as<aod::BigTracksMC>(),
+                                  candidate.index1_as<aod::BigTracksMC>()};
+
+      // Ξcc±± → p± K∓ π± π±
+      //Printf("Checking Ξcc±± → p± K∓ π± π±");
+      //FIXME Be aware that at the moment, you are not considering any resonant channel
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kXiCCPlusPlus, array{+kProton, -kKPlus, +kPiPlus, +kPiPlus}, true, &sign, 3);
+      if (indexRec > -1) {
+        flag = 1 << XiccToXicPi;
+      }
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: `DecayType::XiccToXicPi`

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+      // Check whether the particle is non-prompt (from a b quark).
+      if (flag != 0) {
+        auto particle = particlesMC.iteratorAt(indexRec);
+        origin = (RecoDecay::getMother(particlesMC, particle, 5, true) > -1 ? NonPrompt : Prompt);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: `5` -> `kBottom`
`NonPrompt` -> `OriginType::NonPrompt`

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      //Printf("New gen. candidate");
+      flag = 0;
+      origin = 0;
+      channel = 0;
+      // Xicc → Xic + π+
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kXiCCPlusPlus, array{4232, +kPiPlus}, true)) {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: `4232` -> `pdg::Code::kXiCPlus`

File: Analysis/Core/include/AnalysisCore/HFSelectorCuts.h [hunk 8]
```diff
@@ -331,6 +332,52 @@ static const std::vector<std::string> pTBinLabels = {
...
+  10.0,
+  15.0,
+};
+auto pTBins_v = std::vector<double>{pTBins, pTBins + npTBins + 1};
+
+// default values for the cuts
+//                                            m    CPA  d0Xic  d0Pi  pTXic pTPi
+constexpr double cuts[npTBins][nCutVars] = {{0.5, 0.90, 0.001, 0.001, 2.0, 0.3},  /* 0<pt<0.5 */
+                                            {0.5, 0.90, 0.001, 0.001, 2.0, 0.3},  /* 0.5<pt<1 */
+                                            {0.5, 0.90, 0.001, 0.001, 2.0, 0.3},  /* 1<pt<2   */
```
Comment (vkucera) [inline Analysis/Core/include/AnalysisCore/HFSelectorCuts.h]: Please adjust the spacing in comments for better readability.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        // Match Xic -> pKπ
+        std::vector<int> arrDaughter;
+        RecoDecay::getDaughters(particlesMC, particle, &arrDaughter, array{4232}, 1);
+        auto XicCandMC = particlesMC.iteratorAt(arrDaughter[0]);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Is this necessary? I thought there was a PDG ordering that allows to know at which position which daughter is.

**Review by vkucera — CHANGES_REQUESTED** [review]
General comments:
* Use named numbers.
* Mention the name of the `enum`.
* Avoid `TaskName`.
* Follow naming conventions.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 1 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Hi @vkucera, could you point me to an example? thanks.

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h (hunk 3 above)
Comment (ginnocen) [reply Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:589]: from what I understood this is what is needed in order to match the candidate index with the track index. @rspijkers do you recall how this came up?

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h (hunk 4 above)
Comment (ginnocen) [reply Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h]: not for the moment, removed.

File: Analysis/Tasks/PWGHF/taskXicc.cxx (hunk 2 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/taskXicc.cxx]: Can you specify here what you mean? Gen is usually what identify whether you are referring to a particle.

File: Analysis/Core/include/AnalysisCore/HFSelectorCuts.h (hunk 8 above)
Comment (ginnocen) [reply Analysis/Core/include/AnalysisCore/HFSelectorCuts.h]: done

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 9 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: I dont think here you care about the PDG, it is just trying to retrieve from any of the Xic daughter the index of the Xic.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 7 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: It doesnt like this. See below:

 24   251 |       if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kXiCCPlusPlus, array{pdg::Code::kXiCPlus, +kPiPlus}, true)) {
 25       |                                                                                                                         ^
 26 /home/ginnocen/alice/sw/SOURCES/O2/xicc/0/Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:251:121: error: no matching function for call to 'array(o2::analysis::pdg::Code, int)'

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 6 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: done

File: Analysis/Tasks/PWGHF/taskXicc.cxx (hunk 2 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/taskXicc.cxx]: I would do this in one shot for all hadrons, cause now there are macros that assume this naming convention.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 1 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: hi @vkucera, as discussed we will do it everywhere in one shot. I resolve it for the moment.

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h (hunk 3 above)
Comment (ginnocen) [reply Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:589]: hi @aalkin can you help us with this?

Comment (aalkin) [reply Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:589]: @vkucera @ginnocen this is what provides `.globalIndex()` call for the table. This is a simple enumerator, it is only needed to get a global index for a row, which, in turn is needed if one wants to build a combined index table. Regardless, there is no cost to have this column in table.

File: Analysis/Tasks/PWGHF/taskXicc.cxx (hunk 2 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/taskXicc.cxx]: thanks @vkucera for the clarification. I fixed the name according to the schema of D0 task.

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h (hunk 3 above)
Comment (ginnocen) [reply Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:589]: thanks @aalkin for the clarification.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 9 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: As discussed, this is justified by the ordering that generators preserve when stacking MC particles.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 7 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: thanks, done.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+                                  xicCand.index1_as<aod::BigTracksMC>(),
+                                  xicCand.index2_as<aod::BigTracksMC>(),
+                                  candidate.index1_as<aod::BigTracksMC>()};
+
+      // Ξcc±± → p± K∓ π± π±
+      //Printf("Checking Ξcc±± → p± K∓ π± π±");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kXiCCPlusPlus, array{+kProton, -kKPlus, +kPiPlus, +kPiPlus}, true, &sign, 3);
+      if (indexRec > -1) {
+        flag = 1 << DecayType::XiccToXicPi;
+      }
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Shouldn't depth 2 be used?
Don't you need to also check Ξcc -> Ξc?

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    for (auto& particle : particlesMC) {
+      //Printf("New gen. candidate");
+      flag = 0;
+      origin = 0;
+      // Xicc → Xic + π+
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kXiCCPlusPlus, array{int(pdg::Code::kXiCPlus), +kPiPlus}, true)) {
+        // Match Xic -> pKπ
+        std::vector<int> arrDaughter;
+        RecoDecay::getDaughters(particlesMC, particle, &arrDaughter, array{int(pdg::Code::kXiCPlus)}, 1);
+        auto XicCandMC = particlesMC.iteratorAt(arrDaughter[0]);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: This should be equivalent to `auto XicCandMC = particlesMC.iteratorAt(particle.daughter0());`

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        //Printf("Checking Ξc± → p± K∓ π±");
+        if (RecoDecay::isMatchedMCGen(particlesMC, XicCandMC, int(pdg::Code::kXiCPlus), array{+kProton, -kKPlus, +kPiPlus}, true, &sign)) {
+          flag = sign * (1 << DecayType::XiccToXicPi);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:251]: `sign` should correspond to the one returned by the first `isMatchedMCGen` and the second `isMatchedMCGen` should be called with `acceptAntiParticles = false` to guarantee the same sign of the Ξc daughter.

File: Analysis/Tasks/PWGHF/HFXiccToPKPiPiCandidateSelector.cxx [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    if (hfCandXicc.decayLength() <= cuts->get(pTBin, "decay length")) {
+      return false;
+    }
+
+    // candidate chi2PC
+    if (hfCandXicc.chi2PCA() > cuts->get(pTBin, "chi2PCA")) {
+      return false;
+    }
+    if ((TMath::Abs(hfCandXicc.impactParameter0()) > cuts->get(pTBin, "d0 Xic")) ||
+        (TMath::Abs(hfCandXicc.impactParameter1()) > cuts->get(pTBin, "d0 Pi"))) {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFXiccToPKPiPiCandidateSelector.cxx:81]: `std::abs` is preferable.

File: Analysis/Tasks/PWGHF/taskXicc.cxx [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+#include "Framework/runDataProcessing.h"
+
+/// Ξcc±± analysis task
+struct HfTaskXicc {
+  HistogramRegistry registry{
+    "registry",
+    {{"hptcand", "#Xi^{++}_{cc}-candidates;candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hptprong0", "#Xi^{++}_{cc}-prong candidates;prong 0 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hptprong1", "#Xi^{++}_{cc}-prong candidates;prong 1 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hptprong2", "#Xi^{++}_{cc}-prong candidates;prong 2 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}}}};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskXicc.cxx]: I suppose the title should start with `#Xi^{++}_{cc} candidates` for all.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file HFCandidateCreatorXicc.cxx
+/// \brief Reconstruction of Xiccplusplus candidates
+/// \note Extended from HFCandidateCreator2Prong, HFCandidateCreator3Prong, HFCandidateCreatorX
+///
+/// \author Gian Michele Innocenti <gian.michele.innocenti@cern.ch>, CERN
+/// \author Luigi Dello Stritto <luigi.dello.stritto@cern.ch >, SALERNO
+/// \author Mattia Faggin <mattia.faggin@cern.ch>, University and INFN PADOVA
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:17]: Keep only relevant authors

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+    df2.setBz(magneticField);
+    df2.setPropagateToPCA(b_propdca);
+    df2.setMaxR(d_maxr);
+    df2.setMaxDZIni(d_maxdzini);
+    df2.setMinParamChange(d_minparamchange);
+    df2.setMinRelChi2Change(d_minrelchi2change);
+    df2.setUseAbsDCA(true);
+
+    for (auto& xicCand : xicCands) {
+      if (!(xicCand.hfflag() & 1 << o2::aod::hf_cand_prong3::XicToPKPi)) {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:91]: Missing `DecayType`

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 10 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Yes, I am testing a version.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 11 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx]: Yes, done.

File: Analysis/Tasks/PWGHF/HFXiccToPKPiPiCandidateSelector.cxx (hunk 13 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFXiccToPKPiPiCandidateSelector.cxx:81]: done

File: Analysis/Tasks/PWGHF/taskXicc.cxx (hunk 14 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/taskXicc.cxx]: done

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 15 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:17]: done

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 16 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:91]: done

File: Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx (hunk 12 above)
Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:251]: hi @vkucera I need some clarification on this point, this is why I haven't considered it yet. thanks.

Comment (ginnocen) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorXicc.cxx:251]: Hi @vkucera, I understood the point (finally 👍 ). I will fix it accordingly

**Comment by vkucera** [comment]
Replace by https://github.com/AliceO2Group/AliceO2/pull/6843

**Pull request closed without merging**